File: 2-2.py
#!/usr/bin/env python

import re
import sys

instructions = [insn.strip() for insn in open('2.input').readlines()]

instruction_format = re.compile("(?P<inst>forward|down|up) (?P<oper>[0-9]+)")


class Submarine:
    def __init__(self):
        self.horiz_pos = 0
        self.depth = 0
        self.aim = 0

    def forward(self, steps):
        self.horiz_pos += steps
        self.depth += self.aim * steps
        if self.depth < 0:
            print(f"Illegal direction: forward {steps}")
            sys.exit()

    def down(self, steps):
        self.aim += steps

    def up(self, steps):
        self.aim -= steps

    def __str__(self):
        return f"<{self.__class__.__name__}: ({self.horiz_pos}, {self.depth}, {self.aim})>"


sub = Submarine()

for insn in instructions:
    m = instruction_format.match(insn)
    if m['inst'] == 'forward':
        sub.forward(int(m['oper']))
    elif m['inst'] == 'down':
        sub.down(int(m['oper']))
    elif m['inst'] == 'up':
        sub.up(int(m['oper']))
    else:
        print(f"Illegal instruction: {insn}")
        sys.exit()

print(f"{sub.horiz_pos * sub.depth}")

File: 2-2.rb
#!/usr/bin/env ruby

instructions = File.read('2.input').lines.map(&:strip)

instruction_format = /(?<inst>forward|down|up) (?<oper>[[:digit:]]+)/

class Submarine
  attr_accessor :horiz_pos
  attr_accessor :depth
  attr_accessor :aim

  def initialize
    @horiz_pos = 0
    @depth = 0
    @aim = 0
  end

  def forward(steps)
    @horiz_pos += steps
    @depth += @aim * steps
    if @depth.negative?
      print "Illegal instruction: forward #{steps}"
      exit
    end
  end

  def down(steps)
    @aim += steps
  end

  def up(steps)
    @aim -= steps
  end

  def to_s
    "<#{self.class}: (#{@horiz_pos}, #{@depth}, #{@aim})>"
  end
end

sub = Submarine.new

instructions.each do |insn|
  instruction_format.match(insn) do |m|
    case m['inst']
    when 'forward'
      sub.forward(m['oper'].to_i)
    when 'down'
      sub.down(m['oper'].to_i)
    when 'up'
      sub.up(m['oper'].to_i)
    else
      print "Illegal instruction: #{insn}\n"
      exit
    end
  end
end

print "#{sub.horiz_pos * sub.depth}\n"
